### PR TITLE
BUG: fix construction from read-only non-ns datetime64 numpy array

### LINKED ISSUE #34843: BUG: Series(..) coerces datetime64[non-ns] array differently depending on writable flag
Normally, if we create a Series or DataFrame from a numpy array of non-ns datetime64, and the values are within the ns-range, we convert to datetime64[ns] (in `sanitize_array`). 

However, this does not seem to happen if the array is not writeable:

```
In [9]: arr = np.array(['2020-01-01T00:00:00.000000'], dtype="datetime64[us]") 

In [10]: pd.Series(arr) 
Out[10]: 
0   2020-01-01
dtype: datetime64[ns]

In [11]: arr.flags.writeable = False  

In [12]: pd.Series(arr)  
Out[12]: 
0    2020-01-01 00:00:00
dtype: object
```

I *suppose* this is a bug? (@jbrockmendel ?)

It seems this is due to `tslibs.conversion.ensure_datetime64ns` failing on a read-only buffer, and it started "failing" since pands 0.25

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes #34843

### CODE REVIEW

File: doc/source/whatsnew/v1.1.0.rst [hunk 1]
```diff
@@ -843,6 +843,9 @@ Datetimelike
 - Bug in :meth:`DatetimeIndex.intersection` and :meth:`TimedeltaIndex.intersection` with results not having the correct ``name`` attribute (:issue:`33904`)
 - Bug in :meth:`DatetimeArray.__setitem__`, :meth:`TimedeltaArray.__setitem__`, :meth:`PeriodArray.__setitem__` incorrectly allowing values with ``int64`` dtype to be silently cast (:issue:`33717`)
 - Bug in subtracting :class:`TimedeltaIndex` from :class:`Period` incorrectly raising ``TypeError`` in some cases where it should succeed and ``IncompatibleFrequency`` in some cases where it should raise ``TypeError`` (:issue:`33883`)
+- Bug in constructing a Series or Index from a read-only numpy array with non-ns
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.1.0.rst:846]: `:class:` for Series and Index?  numpy -> NumPy?

File: pandas/tests/base/test_constructors.py [hunk 2]
```diff
@@ -145,3 +158,14 @@ def test_constructor_datetime_outofbound(self, a, klass):
         msg = "Out of bounds"
         with pytest.raises(pd.errors.OutOfBoundsDatetime, match=msg):
             klass(a, dtype="datetime64[ns]")
+
+    def test_constructor_datetime_nonns(self, constructor):
```
Comment (simonjayhawkins) [inline pandas/tests/base/test_constructors.py:162]: add the issue number

Comment (jorisvandenbossche) [reply pandas/tests/base/test_constructors.py:162]: See a few lines below

**Review by TomAugspurger — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Travis failure are the numpy dev warnings

File: pandas/tests/base/test_constructors.py [hunk 3]
```diff
@@ -13,6 +13,19 @@
 from pandas.core.base import NoNewAttributesMixin, PandasObject
 
 
+@pytest.fixture(
```
Comment (jreback) [inline pandas/tests/base/test_constructors.py:16]: @jorisvandenbossche if you hadn’t merged this so quickly i would have objected here

creating this fixture for a single use is not that great

we typically also call this klass and don’t much care about dataframe-dict here

**Comment by jorisvandenbossche** [comment]
> we typically also call this klass 

There are several existing occurrences of a "constructor" fixture, already

> and don’t much care about dataframe-dict here

We actually do, as the test above has an xfail because `DataFrame(dict)` vs `DataFrame(array)` behaves differently (the sanitization of datetime64 values follows a different code path in both cases). 
In this case both work, but so I think it is good to test both.

> creating this fixture for a single use is not that great

Yeah, that is because I first wanted to share it with the other existing test. But because that has an xfail on one of the fixture parameters, making it difficult to share. Once that bug is fixed, it can be shared though.

**Comment by jorisvandenbossche** [comment]
> as the test above has an xfail because DataFrame(dict) vs DataFrame(array) behaves differently (the sanitization of datetime64 values follows a different code path in both cases).

Which is https://github.com/pandas-dev/pandas/issues/26919

**Pull request merged**